File: backend/controller/userController.js
import User from "../models/user.js";
import bcrypt from "bcryptjs";
import asyncHandler from "../utils/asyncHandler.js";

/* ================= GET PROFILE ================= */
export const getProfile = asyncHandler(async (req, res) => {
  const user = await User.findById(req.user._id).select("-password");
  res.json(user);
});

/* ================= SEND OTP ================= */
export const sendOtp = asyncHandler(async (req, res) => {
  const user = await User.findById(req.user._id);

  if (!user) {
    res.status(404);
    throw new Error("User not found");
  }

  // 🔐 Resend limit
  if (user.otpAttempts >= 3) {
    res.status(400);
    throw new Error("Maximum OTP attempts reached");
  }

  const otp = Math.floor(100000 + Math.random() * 900000).toString();

  user.emailOtp = otp;
  user.emailOtpExpire = Date.now() + 5 * 60 * 1000; // 5 min
  user.otpAttempts += 1;

  await user.save();

  // ⚠ For testing (remove in production)
  res.json({
    message: "OTP generated",
    otp,
  });
});

/* ================= UPDATE PROFILE ================= */
export const updateProfile = asyncHandler(async (req, res) => {
  const user = await User.findById(req.user._id);

  if (!user) {
    res.status(404);
    throw new Error("User not found");
  }

  const { name, email, otp, oldPassword, newPassword } = req.body;

  // ✅ Update name
  if (name) {
    user.name = name;
  }

  // ✅ Email change with OTP verification
  if (email && email !== user.email) {
    if (!otp) {
      res.status(400);
      throw new Error("OTP required to change email");
    }

    if (
      user.emailOtp !== otp ||
      !user.emailOtpExpire ||
      user.emailOtpExpire < Date.now()
    ) {
      res.status(400);
      throw new Error("Invalid or expired OTP");
    }

    user.email = email;

    // Reset OTP data
    user.emailOtp = undefined;
    user.emailOtpExpire = undefined;
    user.otpAttempts = 0;
  }

  // ✅ Password change
  if (oldPassword && newPassword) {
    const match = await bcrypt.compare(oldPassword, user.password);

    if (!match) {
      res.status(400);
      throw new Error("Old password incorrect");
    }

    user.password = await bcrypt.hash(newPassword, 10);
  }

  await user.save();

  res.json({
    message: "Profile updated successfully",
    name: user.name,
    email: user.email,
  });
});

File: backend/controller/user_controller.py
"""User profile controller."""
import secrets
import time

from config.db import get_db
from utils.errors import AppError
from utils.security import hash_password, verify_password


async def get_profile(current_user: dict) -> dict:
    db = get_db()
    user = await db.users.find_one({"_id": current_user["_id"]}, {"password": 0})
    if not user:
        raise AppError(404, "User not found")
    return user


async def send_otp(current_user: dict) -> dict:
    db = get_db()
    user = await db.users.find_one({"_id": current_user["_id"]})

    if not user:
        raise AppError(404, "User not found")

    if user.get("otpAttempts", 0) >= 3:
        raise AppError(429, "Maximum OTP attempts reached")

    otp = f"{secrets.randbelow(1_000_000):06d}"

    await db.users.update_one(
        {"_id": user["_id"]},
        {
            "$set": {
                "emailOtp": otp,
                "emailOtpExpire": int(time.time() * 1000) + 5 * 60 * 1000,
            },
            "$inc": {"otpAttempts": 1},
        },
    )

    # Development-only response. Production should deliver the OTP through
    # a trusted email/SMS provider and never expose it in the API response.
    return {"message": "OTP generated", "otp": otp}


async def update_profile(
    current_user: dict,
    name: str | None,
    email: str | None,
    otp: str | None,
    old_password: str | None,
    new_password: str | None,
) -> dict:
    db = get_db()
    user = await db.users.find_one({"_id": current_user["_id"]})

    if not user:
        raise AppError(404, "User not found")

    update: dict = {}

    if name is not None:
        name = name.strip()
        if not name:
            raise AppError(400, "Name cannot be empty")
        update["name"] = name

    if email and email.lower().strip() != user["email"]:
        email = email.lower().strip()
        if not otp:
            raise AppError(400, "OTP required to change email")

        now_ms = int(time.time() * 1000)
        if (
            user.get("emailOtp") != otp
            or not user.get("emailOtpExpire")
            or user["emailOtpExpire"] < now_ms
        ):
            raise AppError(400, "Invalid or expired OTP")

        existing = await db.users.find_one({"email": email, "_id": {"$ne": user["_id"]}})
        if existing:
            raise AppError(409, "Email already in use")

        update["email"] = email
        update["emailOtp"] = None
        update["emailOtpExpire"] = None
        update["otpAttempts"] = 0

    if new_password is not None:
        if not old_password:
            raise AppError(400, "Current password required")
        if len(new_password) < 8:
            raise AppError(400, "New password must be at least 8 characters")
        if not verify_password(old_password, user["password"]):
            raise AppError(400, "Current password incorrect")
        update["password"] = hash_password(new_password)

    if update:
        await db.users.update_one({"_id": user["_id"]}, {"$set": update})

    updated = await db.users.find_one({"_id": user["_id"]}, {"password": 0})

    return {
        "message": "Profile updated successfully",
        "name": updated["name"],
        "email": updated["email"],
    }
